doc/reference/configuration/configuration_reference.rst

```diff
@@ -2861,11 +2861,12 @@ The ``replication`` section defines configuration parameters related to :ref:`re
 roles
 -----
 
-This section describes configuration parameters related to roles.
+This section describes configuration parameters related to application roles.
+An application role is a Lua module that implements specific functions or logic.
 
 .. NOTE::
 
-    Configuration parameters related to roles can be defined in any :ref:`scope <configuration_scopes>`.
+    Configuration parameters related to application roles can be defined in any :ref:`scope <configuration_scopes>`.
```

Comment:
While the configuration merging rules are simple and the same for all the sections, it seems important to highlight them here in context of transforming a cluster configuration (with global/group/replicaset/instance options) to an instance configuration.

It is the point, where behavior is not very intuitive.

First, the rules:

1. Records and maps are deeply merged.
2. Arrays and scalars are *chosen*, not merged. (Note: `any` is a scalar even if its value is a table.)

Instance's option has highest precedence, then replicaset's ones, then group's ones, then globals.

See details about the merging algorithm [here](https://github.com/tarantool/tarantool/blob/8de22969aeeed85428aefb4d608a38c86cd3c3ea/src/box/lua/config/utils/schema.lua#L1299-L1337).

Back to roles.

`roles` is an array. It means that if there is `roles: [foo]` on the global level and `roles: [bar]` on the instance's level, the result is `roles: [bar]`. Not `roles: [foo, bar]`!

It also means that if `roles: [foo, bar, baz]` are defined in globals, in a group or in a replicaset and we need to disable `bar` for a specific instance `i-001` (seems quite usual admin's task), we need to sink the `roles: [foo, bar, baz]` down to all the instances and define `roles: [foo, baz]` for `i-001`.

Something similar with `roles_cfg`: it is `map(string -> any)`. So, `roles_cfg: {foo: {myfoo: 5}}` + `roles_cfg: {bar: {mybar: 6}}` gives `roles_cfg: {foo: {myfoo: 5},  {bar: {mybar: 6}}`, but `roles_cfg: {foo: {myfoo: 5}}` + `roles_cfg: {foo: {myfoo_2: 6}}` gives `roles_cfg: {foo: {myfoo_2: 6}}`. `any` is a scalar.

I would like to discuss how to make things better (I have some ideas around), but now it is important to document constraints and peculiars of the existing implementation.

Reply:
The last bullet in https://docs.d.tarantool.io/en/doc/roles-samples/concepts/configuration/#roles-and-configuration-scopes doesn't correspond tarantool's current behavior.

We consider changing it from the tarantool side. But until it I think that the documentation should match tarantool's behavior.

Leaving the discussion open.

Reply:
Now the description matches tarantool's behavior. Closing.